stdlib/dynarray.mli

```diff
@@ -0,0 +1,418 @@
+
+    It traverses [seq] once and will terminate only if [seq] is finite. *)
+
+val to_seq : 'a t -> 'a Seq.t
+(** [to_seq a] is the sequence of items
+    [get a 0], [get a 1]... [get a (length a - 1)].
+
+    Because sequences are computed on-demand, we have to assume that
+    the array may be modified in the meantime, and give a precise
+    specification for this case below.
+
+    Demanding the [i]-th element of the resulting sequence (which can
+    happen zero, one or several times) will access the [i]-th element
+    of [a] at the time of the demand. The sequence stops if [a] has
+    less than [i] elements at this point.
+*)
+
+val to_seq_rev : 'a t -> 'a Seq.t
+(** [to_seq_rev a] is the sequence of items
+    [get a (l - 1)], [get a (l - 2)]... [get a 0],
+    where [l] is [length a] at the time [to_seq_rev] is invoked.
+
+    Elements that have been removed from the array by the time they
+    are demanded in the sequence are skipped.
+*)
+
+
+(** {1:advanced Advanced topics for performance} *)
+
+(** {2:capacity Backing array, capacity}
+
+    Internally, a dynamic array uses a {b backing array} (a fixed-size
+    array as provided by the {!Array} module) whose length is greater
+    or equal to the length of the dynamic array. We call {b capacity}
+    the length of the backing array.
+
+    The capacity of a dynamic array is relevant in advanced scenarios,
+    when reasoning about the performance of dynamic array programs:
+    {ul
+    {- The memory usage of a dynamic array is proportional to its capacity,
+       rather than its length.}
+    {- Adding elements to the end of a dynamic array may require
+       allocating a new, larger backing array when its length
+       is already equal to its capacity, so there is no room
+       for more elements in the current backing array.}}
+
+    The implementation uses a standard exponential reallocation
+    strategy which guarantees amortized constant-time operation: the
+    total capacity of all backing arrays allocated over the lifetime
+    of a dynamic array is proportional to the total number of elements
+    added or removed.
+    In other words, users need not care about capacity and reallocations,
+    and they will get reasonable behavior by default. However, in some
+    performance-sensitive scenarios the functions below can help control
+    memory usage or guarantee an optimal number of reallocations.
+*)
+
+val ensure_capacity : 'a t -> int -> unit
+(** [ensure_capacity a n] makes sure that [a] has capacity has least [n].
+
+    @raise Invalid_argument if the requested capacity is negative.
+    (We consider that this is a programming error.)
+
+    @raise Failure if the requested capacity is above
+      {!Sys.max_array_length}.
+    (We consider that this is a valid failure mode in some exceptional
+    scenarios. In particular, all functions adding elements to a dynamic
+    array may propagate this exception.)
```

Comment:
I don't think this is a good idea. There's not much you can do at that point and it's inconsistent with `Array` behaviour. Just raise [Invalid_argument] is [n] is not in the range `\[[0];{!Sys.max_array_length}\]`.